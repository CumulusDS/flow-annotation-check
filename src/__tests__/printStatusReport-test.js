'use strict';

/**
 * @flow
 */

import {
  asText,
  asHTMLTable,
  asCSV,
  asJUnit,
} from '../printStatusReport';

import os from 'os';

const BASIC_REPORT = [
  {status: 'flow strict', file: './s.js'},
  {status: 'flow strict-local', file: './sl.js'},
  {status: 'flow', file: './a.js'},
  {status: 'flow weak', file: './b.js'},
  {status: 'no flow', file: './c.js'},
];

const returnTrue = jest.fn();

describe('printStatusReport', () => {
  beforeEach(() => {
    returnTrue.mockReset();
    returnTrue.mockReturnValue(true);
  });

  describe('asText', () => {
    it('should print a simple text report', () => {
      expect(asText(BASIC_REPORT, false, returnTrue)).toMatchSnapshot();
    });

    it('should print a summarized text report', () => {
      expect(asText(BASIC_REPORT, true, returnTrue)).toMatchSnapshot();
    });

    it('should filtered the report based on status', () => {
      asText(BASIC_REPORT, true, returnTrue);
      expect(returnTrue).toHaveBeenCalledTimes(BASIC_REPORT.length);
    });
  });

  describe('asCSV', () => {
    it('should print a simple csv report', () => {
      expect(asCSV(BASIC_REPORT, false, returnTrue)).toMatchSnapshot();
    });

    it('should print a summarized csv report', () => {
      expect(asCSV(BASIC_REPORT, true, returnTrue)).toMatchSnapshot();
    });

    it('should filter the csv report', () => {
      asCSV(BASIC_REPORT, false, returnTrue);
      expect(returnTrue).toHaveBeenCalledTimes(BASIC_REPORT.length);
    });
  });

  describe('asHTMLTable', () => {
    it('should print a simple html-table report', () => {
      expect(asHTMLTable(BASIC_REPORT, false, returnTrue)).toMatchSnapshot();
    });

    it('should print a summarized html-table report', () => {
      expect(asHTMLTable(BASIC_REPORT, true, returnTrue)).toMatchSnapshot();
    });

    it('should filter the html-table report', () => {
      asHTMLTable(BASIC_REPORT, true, returnTrue);
      expect(returnTrue).toHaveBeenCalledTimes(BASIC_REPORT.length);
    });

    it('should print an empty html-table report', () => {
      const report = [];
      expect(asHTMLTable(report, false, returnTrue)).toMatchSnapshot();
    });

    it('should print an html-table report with even percentages', () => {
      const report = [
        {status: 'flow', file: './a.js'},
        {status: 'flow strict', file: './s.js'},
        {status: 'flow weak', file: './b.js'},
        {status: 'no flow', file: './b.js'},
      ];
      expect(asHTMLTable(report, false, returnTrue)).toMatchSnapshot();
    });
  });

  describe('asJUnit', () => {
    beforeEach(() => {
      global.Date = jest.fn(() => ({
        toISOString: () => '-mock date-',
      }));
      // $FlowExpectedError: Overriding for consistency across environments
      os.hostname = jest.fn(() => 'test-host');
    });

    it('should print a jUnit compatible report', () => {
      expect(asJUnit(BASIC_REPORT, returnTrue)).toMatchSnapshot();
    });

    it('should filter the jUnit report', () => {
      asJUnit(BASIC_REPORT, returnTrue);
      expect(returnTrue).toHaveBeenCalledTimes(BASIC_REPORT.length);
    });
  });
});
