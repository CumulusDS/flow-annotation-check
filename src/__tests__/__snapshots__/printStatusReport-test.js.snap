// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printStatusReport asCSV should print a simple csv report 1`] = `
Array [
  "\\"flow strict\\", \\"./s.js\\"",
  "\\"flow strict-local\\", \\"./sl.js\\"",
  "\\"flow\\", \\"./a.js\\"",
  "\\"flow weak\\", \\"./b.js\\"",
  "\\"no flow\\", \\"./c.js\\"",
]
`;

exports[`printStatusReport asCSV should print a summarized csv report 1`] = `
Array [
  "\\"flow strict\\", \\"./s.js\\"",
  "\\"flow strict-local\\", \\"./sl.js\\"",
  "\\"flow\\", \\"./a.js\\"",
  "\\"flow weak\\", \\"./b.js\\"",
  "\\"no flow\\", \\"./c.js\\"",
  "\\"@flow\\", \\"1 (20%)\\"",
  "\\"@flow strict\\", \\"1 (20%)\\"",
  "\\"@flow strict-local\\", \\"1 (20%)\\"",
  "\\"@flow weak\\", \\"1 (20%)\\"",
  "\\"no flow\\", \\"1 (20%)\\"",
  "\\"Total Files\\", \\"5\\"",
]
`;

exports[`printStatusReport asHTMLTable should print a simple html-table report 1`] = `
Array [
  "<table>",
  "<tbody>",
  "<tr data-status=\\"flow strict\\">
<td>flow strict</td>
<td>./s.js</td>
</tr>",
  "<tr data-status=\\"flow strict-local\\">
<td>flow strict-local</td>
<td>./sl.js</td>
</tr>",
  "<tr data-status=\\"flow\\">
<td>flow</td>
<td>./a.js</td>
</tr>",
  "<tr data-status=\\"flow weak\\">
<td>flow weak</td>
<td>./b.js</td>
</tr>",
  "<tr data-status=\\"no flow\\">
<td>no flow</td>
<td>./c.js</td>
</tr>",
  "</tbody>",
  "</table>",
]
`;

exports[`printStatusReport asHTMLTable should print a summarized html-table report 1`] = `
Array [
  "<table>",
  "<tfoot>",
  "<tr><td>@flow</td><td>1 (20%)</td></tr>",
  "<tr><td>@flow strict</td><td>1 (20%)</td></tr>",
  "<tr><td>@flow strict-local</td><td>1 (20%)</td></tr>",
  "<tr><td>@flow weak</td><td>1 (20%)</td></tr>",
  "<tr><td>no flow</td><td>1 (20%)</td></tr>",
  "<tr><td>Total Files</td><td>5</td></tr>",
  "</tfoot>",
  "<tbody>",
  "<tr data-status=\\"flow strict\\">
<td>flow strict</td>
<td>./s.js</td>
</tr>",
  "<tr data-status=\\"flow strict-local\\">
<td>flow strict-local</td>
<td>./sl.js</td>
</tr>",
  "<tr data-status=\\"flow\\">
<td>flow</td>
<td>./a.js</td>
</tr>",
  "<tr data-status=\\"flow weak\\">
<td>flow weak</td>
<td>./b.js</td>
</tr>",
  "<tr data-status=\\"no flow\\">
<td>no flow</td>
<td>./c.js</td>
</tr>",
  "</tbody>",
  "</table>",
]
`;

exports[`printStatusReport asHTMLTable should print an empty html-table report 1`] = `
Array [
  "<table>",
  "<tbody>",
  "</tbody>",
  "</table>",
]
`;

exports[`printStatusReport asHTMLTable should print an html-table report with even percentages 1`] = `
Array [
  "<table>",
  "<tbody>",
  "<tr data-status=\\"flow\\">
<td>flow</td>
<td>./a.js</td>
</tr>",
  "<tr data-status=\\"flow strict\\">
<td>flow strict</td>
<td>./s.js</td>
</tr>",
  "<tr data-status=\\"flow weak\\">
<td>flow weak</td>
<td>./b.js</td>
</tr>",
  "<tr data-status=\\"no flow\\">
<td>no flow</td>
<td>./b.js</td>
</tr>",
  "</tbody>",
  "</table>",
]
`;

exports[`printStatusReport asJUnit should print a jUnit compatible report 1`] = `
Array [
  "<testsuite name=\\"flow-annotation-check\\" timestamp=\\"-mock date-\\" time=\\"0\\" hostname=\\"test-host\\" tests=\\"5\\" failures=\\"2\\" errors=\\"0\\">",
  "<testcase classname=\\"./s.js\\" name=\\"HasFlowStatus\\" time=\\"0\\"><failure type=\\"HasFlowWeakStatus\\">Status is \\"flow strict\\"</failure></testcase>",
  "<testcase classname=\\"./sl.js\\" name=\\"HasFlowStatus\\" time=\\"0\\"><failure type=\\"HasFlowWeakStatus\\">Status is \\"flow strict-local\\"</failure></testcase>",
  "<testcase classname=\\"./a.js\\" name=\\"HasFlowStatus\\" time=\\"0\\" />",
  "<testcase classname=\\"./b.js\\" name=\\"HasFlowStatus\\" time=\\"0\\"><failure type=\\"HasFlowWeakStatus\\">Status is \\"flow weak\\"</failure></testcase>",
  "<testcase classname=\\"./c.js\\" name=\\"HasFlowStatus\\" time=\\"0\\"><failure type=\\"HasNoneStatus\\">Status is \\"no flow\\"</failure></testcase>",
  "</testsuite>",
]
`;

exports[`printStatusReport asText should print a simple text report 1`] = `
Array [
  "flow strict	./s.js",
  "flow strict-local	./sl.js",
  "flow	./a.js",
  "flow weak	./b.js",
  "no flow	./c.js",
]
`;

exports[`printStatusReport asText should print a summarized text report 1`] = `
Array [
  "flow strict	./s.js",
  "flow strict-local	./sl.js",
  "flow	./a.js",
  "flow weak	./b.js",
  "no flow	./c.js",
  "@flow 1 (20%)",
  "@flow strict 1 (20%)",
  "@flow strict-local 1 (20%)",
  "@flow weak 1 (20%)",
  "no flow 1 (20%)",
  "Total Files 5",
]
`;
